compiler: improve string_utils and diagnostics output

chqrlie opened the pull request:
string_buffer:
* remove `string_buffer.Buf.add_line()`
* add `string_buffer.Buf.vprintf()`

string_utils:
* remove `string_utils.count_tabs()`
* add `string_utils.toLower()`
* fix UB on empty string in `stripNewLine()`
* handle `\r\n` in `stripNewLine()`
* return destination pointer in `toLower()`, `toUpper()`,  `stripNewline()`

utf8:
* add `utf8_sync()` to find character boundaries

parser_utils:
* return offset past the end in `getTokenEnd()`

diagnostics:
* improve range handling
* display line fragment for long lines
* detect and ignore region spanning multiple lines
* handle tabs anywhere on source line

bvdberg commented:
It took some time to compare the diagnostics.c2. I understand what you did. Mostly pulling the end loc stuff in a smaller scope. The code for skipping columns in long lines seems weird. loc_col is modified (loc_col -= skip) but never used after that.

I think i counted the tabs, to get the '^' at the right place as well, since the tabsize may differ per system, i'm not sure it still  works currently..

chqrlie commented:
> It took some time to compare the diagnostics.c2. I understand what you did. Mostly pulling the end loc stuff in a smaller scope. The code for skipping columns in long lines seems weird. loc_col is modified (loc_col -= skip) but never used after that.

OOPS! that's a bug, `if (col == loc.column) c = '^';` must be changed to `if (col == loc_col) c = '^';`

I added more comments to explain the methodology.

> I think i counted the tabs, to get the '^' at the right place as well, since the tabsize may differ per system, i'm not sure it still works currently..

Yes, you were counting the initial TABs for this purpose, but that is not sufficient if the source line contains TABs in the middle of the line, for example to align columns in an array initializer.  I changed the code to output a TAB in every position where there is a TAB in the source line above (or the visible portion thereof).  This does not account for TABs inside the range, but this should be rare.  I personally never use TABs (except in Makefile) but rejecting them seems unwarranted.

Another source of potential misalignment is unicode characters that may have a zero width or a double width. Such characters could be present in the source line inside strings or comments. I wrote code to support that in **qemacs**, but it is a low priority.

bvdberg commented:
I found 2 issues:
- one is utf8.utf8_sync, this is an anti-pattern, just name the function utf8_sync -> sync in module utf8 (no it wont clash ;) )
- the second is that the ^ is now broken, check:
const char* str = "abc"

It gives an error:
dd.c2:3:24: error: expected ';'
const char* str = "abc"
                           <- empty line, missing ^

On the master branch it gives:

dd.c2:3:24: error: expected ';'
const char* str = "abc"
                                        ^     (it should point to 1 column to the right of ")

I'll see if i can spot the issue, it's not a long line, so the code should basically do the same as before

chqrlie commented:
I fixed both issues.  renamed `utf8_sync` as `sync` in module `utf8`... importing it inline is not recommended.

The ^ issue was an off by one check when the error is exactly at the end of a line (with ot without a newline).
